### PR TITLE
SDIT-1518: ✨ Initial stab at subject access request report

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/model/response/PrisonApiResponses.kt [hunk 1]
```diff
@@ -12,11 +10,4 @@ data class MovementResponse(
   val movementReasonCode: String?,
 )
 
-data class RestrictedPatient(
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/model/response/PrisonApiResponses.kt:15]: not really related but found it was unused so removed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/model/response/SubjectAccessRequestContent.kt [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package uk.gov.justice.digital.hmpps.hmppsrestrictedpatientsapi.model.response
+
+import io.swagger.v3.oas.annotations.media.Schema
+import java.time.LocalDateTime
+
+data class SubjectAccessRequestContent(
+  @Schema(description = "The content of the subject access request response")
+  val content: RestrictedPatientContent,
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/model/response/SubjectAccessRequestContent.kt:8]: need a `content` at top level in the response

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/controllers/SubjectAccessRequestController.kt [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        ?: ResponseEntity.status(HttpStatus.NO_CONTENT)
+          .body(
+            ErrorResponse(
+              status = HttpStatus.NO_CONTENT,
+              userMessage = "Request successfully processed - no content found",
+              developerMessage = "Request successfully processed - no content found",
+            ),
+          )
+    }
+    return ResponseEntity.status(209).body(
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/controllers/SubjectAccessRequestController.kt:100]: not really clear what should happen if both `prn` and `crn` supplied, but I've gone for a `209` if `prn` is missing 🤷

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/model/response/SubjectAccessRequestContent.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+)
+
+data class RestrictedPatientContent(
+  @Schema(description = "The prisoner number (Nomis ID)", example = "A1234AA")
+  val prisonerNumber: String,
+  @Schema(
+    description = "Prison where the offender is supported by POM.  This is from AGENCY_LOCATIONS.AGY_LOC_ID in NOMIS.",
+    example = "LEI",
+  )
+  val supportingPrisonId: String,
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/model/response/SubjectAccessRequestContent.kt:18]: I've opted for only supplying the data we've got for now and haven't done lookups to Prison API.  Will add them later if needed

Comment (petergphillips) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/model/response/SubjectAccessRequestContent.kt:18]: also the supporting prison isn't held by Prison API so we return it - can be different from the discharge prison (which will be held in NOMIS as the last movement from location)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/controllers/SubjectAccessRequestController.kt [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            schema = Schema(implementation = SubjectAccessRequestContent::class),
+          ),
+        ],
+      ),
+      ApiResponse(
+        responseCode = "204",
+        description = "Request successfully processed - no content found",
+      ),
+      ApiResponse(
+        responseCode = "209",
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/controllers/SubjectAccessRequestController.kt:53]: 209? 404 surely 😆 (yes I know this wasn't your call)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/controllers/SubjectAccessRequestController.kt [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    @RequestParam(name = "prn", required = false)
+    @Parameter(description = "NOMIS Prison Reference Number")
+    prn: String?,
+    @RequestParam(name = "crn", required = false)
+    @Parameter(description = "nDelius Case Reference Number")
+    crn: String?,
+    @RequestParam(value = "fromDate", required = false)
+    @Parameter(description = "Optional parameter denoting minimum date of event occurrence which should be returned in the response")
+    fromDate: LocalDate?,
+    @RequestParam(value = "toDate", required = false)
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/controllers/SubjectAccessRequestController.kt:84]: Don't think you need required here?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/controllers/SubjectAccessRequestController.kt (hunk 3 above)
Comment (mikehalmamoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/controllers/SubjectAccessRequestController.kt:100]: Feels like a 4xx but response codes seem odd for SAR anyway

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/controllers/SubjectAccessRequestController.kt (hunk 5 above)
Comment (petergphillips) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/controllers/SubjectAccessRequestController.kt:53]: `418` seems more appropriate

**Pull request merged**